### ISSUE TITLE
Training Paradigm: Task-Conditioned Pretraining and Zero-Shot Prediction (EveryQuery-style)

### ISSUE THREAD

**Issue by stepanyanhayk** [McDermottHealthAI/MedRAP#38]
## Motivation

The current MedRAP setup is **single-task**: train on mortality in 24h, evaluate on mortality in 24h. If we want to predict a different outcome (e.g., sepsis onset, readmission in 30 days), we need to collect new labels and retrain.

**EveryQuery (EQ)** (arXiv:2603.07900) proposes a fundamentally different training paradigm: instead of fixing a single prediction task, **randomly sample (patient, task_specification, time_window) triples during pretraining** and train the model to predict the outcome defined by the task spec given the patient history up to the time window. At inference, **any** clinical prediction task expressible in the query language can be answered **zero-shot** — no fine-tuning, no new labels.

EQ demonstrated:
- Outperforms autoregressive baselines on **82% of 39 randomly sampled MIMIC-IV tasks**
- Mean AUC improvement of **+0.16** over AR baselines
- Particularly strong on **rare events** (where AR trajectory generation fails due to error accumulation)

**Why this matters for MedRAP specifically:** The retrieval step in MedRAP already conditions on a patient query. EQ suggests we should also condition on a *task query* — the model should know *what* it is predicting when it retrieves documents and fuses them. A model predicting "sepsis" should retrieve sepsis-relevant papers; a model predicting "readmission" should retrieve different ones. Currently MedRAP retrieves based only on patient history, not on the prediction target.

This combines naturally with the existing marginalized retrieval infrastructure (`MarginalizedRetrievalLoss`, `differentiable_retrieval_scores`) — the retrieval can be conditioned on both patient state and task specification.

### Core idea
Extend the input to the model to include a **task specification** alongside patient history:

```
Input:  (patient_history, task_spec)  →  query  →  retrieve task-relevant docs  →  predict
```

where `task_spec` encodes:
- **Outcome code(s):** What clinical event to predict (ICD code, lab flag, etc.)
- **Time window:** How far in the future (e.g., 24h, 7d, 30d)

The query projector then produces a query conditioned on *both* the patient state *and* the task spec. Retrieved documents are relevant to both the patient's clinical profile and the prediction task.

### Pretraining procedure (EveryQuery-style)
During pretraining, for each batch:
1. Sample a patient history up to a random cutoff time
2. Sample a random task spec from a task library (e.g., 500 clinical outcome codes × 5 time windows = 2,500 combinations)
3. Look up whether the outcome occurred in the patient's future (label)
4. Train with BCE loss on the prediction

At inference, plug in any task spec → zero-shot prediction without fine-tuning.

### Small — Fixed task library, one-hot task conditioning
- Task library: 20 clinical outcomes × 3 time windows (24h, 7d, 30d) = 60 task specs
- Task encoding: one-hot vector (dim=60) projected via linear layer → D=128, concatenated to the patient query vector before retrieval
- Encoder: existing `TransformerPatientEncoder` (D=128, 2L) — no change
- Pretraining: train on 60 tasks uniformly sampled; evaluate zero-shot on held-out tasks
- Query projector: `SequenceMeanQueryProjector` → concatenate task embedding → linear → D=1024
- ~+8K parameters (task embedding table + projection)

### Medium — Learned task embeddings, cross-attention conditioning
- Task library: 200 clinical outcomes × 5 time windows = 1,000 task specs
- Task encoding: learned embedding per task spec (dim=256), cross-attended with patient sequence before mean-pooling
- Cross-attention: 1 layer, patient sequence as query, task embedding as key/value → contextualised patient rep conditioned on task
- Encoder: D=256, 4 layers, 8 heads (pre-norm)
- Retrieval: query projected from task-conditioned patient rep → retrieves *task-relevant* medical literature
- ~+2M parameters

### Large — Natural language task queries, full task-conditioned retrieval
- Task encoding: encode task spec as natural language ("predict mortality within 24 hours in ICU patients") using a frozen small LLM (e.g., Qwen3-0.6B encoder, same as current retrieval encoder)
- Cross-attention: 2-layer transformer cross-attention between task LLM embedding and patient sequence
- Retrieval: query = task-conditioned patient rep → retrieves docs relevant to *both* patient AND task (e.g., if task = "sepsis 24h", retrieves sepsis management guidelines)
- Encoder: D=512, 6 layers, 16 heads
- Enables strong zero-shot generalisation: task specs never seen during pretraining can still leverage the NL encoding
- ~+10M parameters (excluding frozen LLM)

## Key design questions
- What is the right task library for pretraining? EQ uses randomly sampled ICD codes; we should stratify by frequency (ensure rare outcomes are covered).
- Should the task spec condition the *retrieval query* (what documents to retrieve), the *fusion* (how to attend to retrieved docs), or both?
- Can the existing `marginalized_retrieval` loss be extended to also marginalise over task specs at pretraining time?
- How does zero-shot performance on unseen tasks compare to single-task fine-tuning? This is the core empirical question.

## Relation to current codebase
- `model.py` — add `task_spec` as optional input to `RetrievalAugmentedModel.forward()` (defaults to None for backwards compatibility)
- `query_projection.py` — add `TaskConditionedQueryProjector` that takes both `patient_state` and `task_embedding`
- `datamodule.py` — extend `MEDSTorchBatch`-based loading to include multi-task label sampling during pretraining
- `task.py` — add `MultiTaskBinaryClassificationTask` that samples from a task library
- `conf/` — add `training/task/multi_task_small.yaml`, `medium`, `large`
- `scripts/` — add `pretrain_task_conditioned.sh`

## References
- EveryQuery paper: arXiv:2603.07900 (read the paper for exact query language design and pretraining details)
- Key result: "EveryQuery outperforms on 82% of tasks, mean AUC +0.16, especially on rare events"
- Limitation to be aware of: disjunctive queries ("outcome A OR B") currently fail in EQ; design task specs as conjunctive only
